source/common/http/http1/codec_impl.cc

```diff
@@ -192,13 +194,18 @@ void StreamEncoderImpl::readDisable(bool disable) { connection_.readDisable(disa
 uint32_t StreamEncoderImpl::bufferLimit() { return connection_.bufferLimit(); }
 
 static const char RESPONSE_PREFIX[] = "HTTP/1.1 ";
+static const char HTTP_10_RESPONSE_PREFIX[] = "HTTP/1.0 ";
 
 void ResponseStreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_stream) {
   started_response_ = true;
   uint64_t numeric_status = Utility::getResponseStatus(headers);
 
   connection_.reserveBuffer(4096);
-  connection_.copyToBuffer(RESPONSE_PREFIX, sizeof(RESPONSE_PREFIX) - 1);
+  if (connection_.protocol() == Protocol::Http10 && connection_.supports_http_10()) {
```

Comment:
We should send response with `HTTP/1.1` version even if request is HTTP/1.0, see [RFC7230, sec. 2.6 Protocol Versioning](https://tools.ietf.org/html/rfc7230#section-2.6).

Reply:
If I read that correctly it says you should reply with the version you are attempting to adhere to. We are attempting to respond with HTTP/1.0.  If you can find me something in the HTTP/1.0 spec which says it must accept HTTP/1.1 responses I'd be more inclined to not downgrade.

Reply:
It's a HTTP/1.0 response, but emitted by HTTP/1.1 server... Basically, it's saying "FYI, I can speak HTTP/1.1 even though you've requested HTTP/1.0":

```
   The HTTP version number consists of two decimal digits separated by a
   "." (period or decimal point).  The first digit ("major version")
   indicates the HTTP messaging syntax, whereas the second digit ("minor
   version") indicates the highest minor version within that major
   version to which the sender is conformant and able to understand for
   future communication.  The minor version advertises the sender's
   communication capabilities even when the sender is only using a
   backwards-compatible subset of the protocol, thereby letting the
   recipient know that more advanced features can be used in response
   (by servers) or in future requests (by clients).
```

HTTP/1.0 clients should accept all HTTP/1.x responses:

```
   When an HTTP message is received with a major version number that the
   recipient implements, but a higher minor version number than what the
   recipient implements, the recipient SHOULD process the message as if
   it were in the highest minor version within that major version to
   which the recipient is conformant.  A recipient can assume that a
   message with a higher minor version, when sent to a recipient that
   has not yet indicated support for that higher version, is
   sufficiently backwards-compatible to be safely processed by any
   implementation of the same major version.
```

assuming that response doesn't include any features requiring newer minor version:

```
   When an HTTP/1.1 message is sent to an HTTP/1.0 recipient [RFC1945]
   or a recipient whose version is unknown, the HTTP/1.1 message is
   constructed such that it can be interpreted as a valid HTTP/1.0
   message if all of the newer features are ignored.  This specification
   places recipient-version requirements on some new features so that a
   conformant sender will only use compatible features until it has
   determined, through configuration or the receipt of a message, that
   the recipient supports HTTP/1.1.
```

i.e. no chunking, etc. is allowed when replying to HTTP/1.0 client.

This is definitely legal behavior (again, NGINX), but I'm fine if you want to keep it as HTTP/1.0.

Reply:
Ah, that's both useful and clear, thanks for the detailed response.  Given you said you're OK either way, I'm inclined to explicitly respond with HTTP/1.0 both because it makes debugging easier and because that's the behavior I'm used to seeing so I'm more confident it will work for the clients I'm implementing this for.